core/server/common/src/main/java/alluxio/master/BackupManager.java

```diff
@@ -83,27 +215,130 @@ public void backup(OutputStream os) throws IOException {
+        } catch (Exception exc) {
+          // Store exception.
+          readerTaskFuture.setException(exc);
+        } finally {
+          readerTaskFuture.set(null);
         }
-        try (JournalContext jc = master.createJournalContext()) {
-          jc.append(entry);
-          count++;
+      });
+      // Will shutdown as soon as the pending tasks are completed.
+      backupReaderExecutor.shutdown();
+
+      // Create applier task.
+      backupApplierExecutor.submit(() -> {
+        try {
+          while (!readerTaskFuture.isDone() || journalEntryQueue.size() > 0) {
+            // Drain current elements.
+            // Draining entries makes it possible to allow writes while current ones are
+            // being applied.
+            List<JournalEntry> drainedEntries = new LinkedList<>();
+            if (0 == journalEntryQueue.drainTo(drainedEntries)) {
+              // No elements at the moment. Fall back to polling.
+              JournalEntry entry = journalEntryQueue.poll(10, TimeUnit.MILLISECONDS);
+              if (entry == null) {
+                // No entry yet.
+                continue;
+              }
+              drainedEntries.add(entry);
+            }
+            // Apply entries.
+            for (JournalEntry entry : drainedEntries) {
+              Master master = mastersByName.get(JournalEntryAssociation.getMasterForEntry(entry));
+              master.applyAndJournal(masterJCMap.get(master), entry);
+            }
+          }
+        } catch (InterruptedException ie) {
+          // Store exception.
+          applierTaskFuture.setException(ie);
+          // Continue interrupt chain.
+          Thread.currentThread().interrupt();
+          throw new RuntimeException("Thread interrupted while applying backup content.", ie);
+        } catch (Exception exc) {
+          // Store exception.
+          applierTaskFuture.setException(exc);
+        } finally {
+          applierTaskFuture.set(null);
+        }
+      });
+      // Will shutdown as soon as the pending tasks are completed.
+      backupApplierExecutor.shutdown();
+
+      // Wait until backup initialization is done.
+      try {
+        // Wait for backup reader.
+        readerTaskFuture.get();
+        // Wait for backup applier.
+        applierTaskFuture.get();
+      } catch (InterruptedException ie) {
+        // Continue interrupt chain.
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("Thread interrupted while waiting for backup threads.", ie);
+      } catch (ExecutionException ee) {
+        Throwable cause = ee.getCause();
+        if (cause instanceof IOException) {
+          throw (IOException) cause;
+        } else {
+          throw new IOException(cause);
+        }
+      } finally {
+        // Close journal contexts to ensure applied entries are flushed.
+        for (JournalContext journalContext : masterJCMap.values()) {
+          journalContext.close();
```

Comment:
will this work if the metadata size is large? We can't afford to buffer the full backup in memory. I think we might need to write it in chunks instead

Reply:
Journal writer has a dedicated thread that continuously writes. So it'll never be all of the entries of backup. We can have intermittent flushes here at each drain time, if you think it's still necessary with dedicated thread on journal flushing side.

Reply:
I see, it will only be an issue if the tree crawling threads produce inodes faster than the writer thread can write them to the backup.

Perhaps `JournalContext` is the wrong tool to use here. It is designed to support many threads writing to the journal concurrently, but in this situation we have only a single thread. Perhaps we could replace it with a simple bounded queue (to prevent unlimited memory usage) plus a regular `JournalWriter` instead of `AsyncJournalWriter`?

Reply:
I think doing an intermittent flushes per each drained entry list is acceptable enough. For 5M files I didn't see any perf change. For bigger loads, it shouldn't be any worse. The fix is to open and close contexts per each round of draining from queue. This will make sure the amount of memory we use is proportional to the configured buffer size.